File: src/main/java/net/atos/zac/notities/NotitieService.java
/*
 * SPDX-FileCopyrightText: 2021 Atos
 * SPDX-License-Identifier: EUPL-1.2+
 */

package net.atos.zac.notities;

import static net.atos.zac.notities.model.Notitie.ZAAK_UUID;
import static net.atos.zac.util.ValidationUtil.valideerObject;

import java.util.List;
import java.util.UUID;

import jakarta.enterprise.context.ApplicationScoped;
import jakarta.persistence.EntityManager;
import jakarta.persistence.PersistenceContext;
import jakarta.persistence.criteria.CriteriaBuilder;
import jakarta.persistence.criteria.CriteriaQuery;
import jakarta.persistence.criteria.Root;
import jakarta.transaction.Transactional;

import net.atos.zac.notities.model.Notitie;

@ApplicationScoped
@Transactional
public class NotitieService {

    @PersistenceContext(unitName = "ZaakafhandelcomponentPU")
    private EntityManager entityManager;

    public Notitie createNotitie(final Notitie notitie) {
        valideerObject(notitie);
        entityManager.persist(notitie);
        return notitie;
    }

    public List<Notitie> listNotitiesForZaak(final UUID zaakUUID) {
        final CriteriaBuilder builder = entityManager.getCriteriaBuilder();
        final CriteriaQuery<Notitie> query = builder.createQuery(Notitie.class);
        final Root<Notitie> root = query.from(Notitie.class);
        query.select(root).where(builder.equal(root.get(ZAAK_UUID), zaakUUID));
        return entityManager.createQuery(query).getResultList();
    }

    public Notitie updateNotitie(final Notitie notitie) {
        valideerObject(notitie);
        return entityManager.merge(notitie);
    }

    public void deleteNotitie(final Long notitieId) {
        final Notitie notitie = entityManager.find(Notitie.class, notitieId);
        entityManager.remove(notitie);
    }
}

File: src/main/kotlin/nl/info/zac/notities/NotitieService.kt
/*
 * SPDX-FileCopyrightText: 2021 Atos, 2025 INFO.nl
 * SPDX-License-Identifier: EUPL-1.2+
 */
package nl.info.zac.notities

import jakarta.enterprise.context.ApplicationScoped
import jakarta.inject.Inject
import jakarta.persistence.EntityManager
import jakarta.transaction.Transactional
import jakarta.transaction.Transactional.TxType.REQUIRED
import jakarta.transaction.Transactional.TxType.SUPPORTS
import net.atos.zac.util.ValidationUtil
import nl.info.zac.notities.model.Notitie
import nl.info.zac.notities.model.Notitie.Companion.ZAAK_UUID_FIELD
import nl.info.zac.util.AllOpen
import nl.info.zac.util.NoArgConstructor
import java.util.UUID

@ApplicationScoped
@Transactional(SUPPORTS)
@NoArgConstructor
@AllOpen
class NotitieService @Inject constructor(
    private val entityManager: EntityManager
) {
    @Transactional(REQUIRED)
    fun createNotitie(notitie: Notitie): Notitie {
        ValidationUtil.valideerObject(notitie)
        entityManager.persist(notitie)
        return notitie
    }

    fun listNotitiesForZaak(zaakUUID: UUID): List<Notitie> {
        val builder = entityManager.criteriaBuilder
        val query = builder.createQuery(Notitie::class.java)
        val root = query.from(Notitie::class.java)
        query.select(root).where(builder.equal(root.get<Any>(ZAAK_UUID_FIELD), zaakUUID))
        return entityManager.createQuery(query).getResultList()
    }

    @Transactional(REQUIRED)
    fun updateNotitie(notitie: Notitie): Notitie {
        ValidationUtil.valideerObject(notitie)
        return entityManager.merge(notitie)
    }

    @Transactional(REQUIRED)
    fun deleteNotitie(notitieId: Long) {
        val notitie = entityManager.find(Notitie::class.java, notitieId)
        entityManager.remove(notitie)
    }
}
